fdbserver/DataDistributionTracker.actor.cpp

```diff
@@ -154,65 +151,61 @@ ACTOR Future<Void> trackShardBytes(
-						bounds.max.bytesPerKSecond = SERVER_KNOBS->SHARD_MIN_BYTES_PER_KSEC;
-						bounds.min.bytesPerKSecond = 0;
-						bounds.permittedError.bytesPerKSecond = bounds.max.bytesPerKSecond / 4;
-					} else
-						ASSERT( false );
-
-				} else {
-					bounds.max.bytes = -1;
-					bounds.min.bytes = -1;
-					bounds.permittedError.bytes = -1;
+			ShardSizeBounds bounds;
+			if( shardSize->get().present() ) {
+				auto bytes = shardSize->get().get().bytes;
+				auto bandwidthStatus = getBandwidthStatus( shardSize->get().get() );
+				bounds.max.bytes = std::max( int64_t(bytes * 1.1), (int64_t)SERVER_KNOBS->MIN_SHARD_BYTES );
+				bounds.min.bytes = std::min( int64_t(bytes * 0.9), std::max(int64_t(bytes - (SERVER_KNOBS->MIN_SHARD_BYTES * 0.1)), (int64_t)0) );
+				bounds.permittedError.bytes = bytes * 0.1;
+				if( bandwidthStatus == BandwidthStatusNormal ) {   // Not high or low
+					bounds.max.bytesPerKSecond = SERVER_KNOBS->SHARD_MAX_BYTES_PER_KSEC;
+					bounds.min.bytesPerKSecond = SERVER_KNOBS->SHARD_MIN_BYTES_PER_KSEC;
+					bounds.permittedError.bytesPerKSecond = bounds.min.bytesPerKSecond / 4;
+				} else if( bandwidthStatus == BandwidthStatusHigh ) { // > 10MB/sec for 100MB shard, proportionally lower for smaller shard, > 200KB/sec no matter what
 					bounds.max.bytesPerKSecond = bounds.max.infinity;
+					bounds.min.bytesPerKSecond = SERVER_KNOBS->SHARD_MAX_BYTES_PER_KSEC;
+					bounds.permittedError.bytesPerKSecond = bounds.min.bytesPerKSecond / 4;
+				} else if( bandwidthStatus == BandwidthStatusLow ) {  // < 10KB/sec
+					bounds.max.bytesPerKSecond = SERVER_KNOBS->SHARD_MIN_BYTES_PER_KSEC;
 					bounds.min.bytesPerKSecond = 0;
-					bounds.permittedError.bytesPerKSecond = bounds.permittedError.infinity;
-				}
-
-				bounds.max.iosPerKSecond = bounds.max.infinity;
-				bounds.min.iosPerKSecond = 0;
-				bounds.permittedError.iosPerKSecond = bounds.permittedError.infinity;
-
-				StorageMetrics metrics = wait( tr.waitStorageMetrics( keys, bounds.min, bounds.max, bounds.permittedError, CLIENT_KNOBS->STORAGE_METRICS_SHARD_LIMIT ) );
-
-				/*TraceEvent("ShardSizeUpdate")
-					.detail("Keys", printable(keys))
-					.detail("UpdatedSize", metrics.metrics.bytes)
-					.detail("Bandwidth", metrics.metrics.bytesPerKSecond)
-					.detail("BandwithStatus", getBandwidthStatus(metrics))
-					.detail("BytesLower", bounds.min.bytes)
-					.detail("BytesUpper", bounds.max.bytes)
-					.detail("BandwidthLower", bounds.min.bytesPerKSecond)
-					.detail("BandwidthUpper", bounds.max.bytesPerKSecond)
-					.detail("ShardSizePresent", shardSize->get().present())
-					.detail("OldShardSize", shardSize->get().present() ? shardSize->get().get().metrics.bytes : 0 )
-					.detail("TrackerID", trackerID);*/
-
-				if( shardSize->get().present() && addToSizeEstimate )
-					self->dbSizeEstimate->set( self->dbSizeEstimate->get() + metrics.bytes - shardSize->get().get().bytes );
-
-				shardSize->set( metrics );
-			} catch( Error &e ) {
```

Comment:
It's not obvious to me why this can be removed. I see that there is one wait on a call to `waitStorageMetrics`. Is the idea that it can't throw any errors where we would retry them?